streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -16,171 +16,256 @@
+     */
+    boolean updateNewAndRestoringTasks() {
+        final List<Task> restoringTasks = new LinkedList<>();
+        for (final Task task : tasks.values()) {
+            task.initializeIfNeeded();
+            if (task.state() == RESTORING) {
+                restoringTasks.add(task);
+            }
+        }
 
-        for (final TaskId taskId : suspendedTasks) {
-            final Set<TopicPartition> partitions = assignedActiveTasks.get(taskId);
-            try {
-                if (!active.maybeResumeSuspendedTask(taskId, partitions)) {
-                    // recreate if resuming the suspended task failed because the associated partitions changed
-                    addedActiveTasks.put(taskId, partitions);
+        boolean allRunning = true;
+        if (!restoringTasks.isEmpty()) {
+            changelogReader.restore();
+            final Set<TopicPartition> restored = changelogReader.completedChangelogs();
+            for (final Task task : restoringTasks) {
+                if (restored.containsAll(task.changelogPartitions())) {
+                    task.startRunning();
+                } else {
+                    // we found a restoring task that isn't done restoring, which is evidence that
+                    // not all tasks are running
+                    allRunning = false;
                 }
-            } catch (final StreamsException e) {
-                log.error("Failed to resume a suspended active task {} due to the following error:", taskId, e);
-                throw e;
             }
         }
+
+        // Note: if we wanted to process ready tasks while others are still restoring, all we have to do
+        // is delete this conditional and run the loop always.
+        if (allRunning) {
+            for (final Task task : tasks.values()) {
+                // TODO, can we make StandbyTasks partitions always empty (since they don't process any inputs)?
+                // If so, we can simplify this logic here, as the resume would be a no-op.
+                if (task.isActive()) {
+                    consumer.resume(task.inputPartitions());
+                }
+            }
+        }
+        return allRunning;
     }
 
-    private void addNewActiveTasks(final Map<TaskId, Set<TopicPartition>> newActiveTasks) {
-        log.debug("New active tasks to be created: {}", newActiveTasks);
+    /**
+     * Similar to shutdownTasksAndState, however does not close the task managers, in the hope that
+     * soon the tasks will be assigned again.
+     * @throws TaskMigratedException if the task producer got fenced (EOS only)
+     */
+    void handleRevocation(final Collection<TopicPartition> revokedPartitions) {
+        final Set<TaskId> revokedTasks = new HashSet<>();
+
+        for (final Task task : tasks.values()) {
+            for (final TopicPartition partition : revokedPartitions) {
+                if (task.inputPartitions().contains(partition)) {
```

Comment:
This is not introduced by this PR but I'm wondering if we should check that ALL partitions of this task is included? Assuming we are bug-free this should be fine too but we can be stricter.

Reply:
It seems like it would be an assignor bug if only some of the partitions for a task are revoked (or likewise if only some of the partitions for a task are assigned). Not a counter-argument; it may well be worthwhile to detect up front that we may be executing a different topology than the one the assignor used, or that the assignor has screwed up the assignment in some way.